Python/ast_opt.c

```diff
@@ -41,20 +43,27 @@ typedef struct {
     } while(0)
 
 
-static ControlFlowInFinallyContext
-overwrite_state(_PyASTOptimizeState *state, bool finally, bool funcdef, bool loop)
+static int
+push_cf_context(_PyASTOptimizeState *state, stmt_ty node, bool finally, bool funcdef, bool loop)
 {
-    ControlFlowInFinallyContext saved = state->cf_finally;
-    state->cf_finally.in_finally = finally;
-    state->cf_finally.in_funcdef = funcdef;
-    state->cf_finally.in_loop = loop;
-    return saved;
+    if (state->cf_finally_next == CO_MAXBLOCKS) {
```

Comment:
This is a slightly different use of `CO_MAXBLOCKS`.
Currently, the compiler pushes a "block" for loops (`while` and `for`), `with` and try-finallys.
This will push a context for each function definition, loop or finally, but not `with` statements.

For consistency, how about pushing a fake context for `with` and try-except here and dropping the check in `compiler.c`?
To avoid breaking existing code, we'll need to increase `CO_MAXBLOCKS` to, say, 32.

Reply:
Adding context for with and try wouldn't make it consistent with compiler.c, because compiler.c doesn't push a block for a function definition and here we do.

I think we should use a different limit here than `CO_MAXBLOCKS`. The number of nested functions is limited by the parser's MAXINDENT (100).

Reply:
Adding contexts for `with` and `try` would mean that there is no case where compiler.c can overflow its stack, without it being caught here first.

 Upping the limit to match the parser sounds reasonable, although it might make sense to reduce `MAXINDENT`. There is no point in being able to parse code that we can't then compile.

Reply:
> There is no point in being able to parse code that we can't then compile.

There is no problem compiling a lot of nested functions. A function is not a static block for the compiler.

```
>>> N = 120
... 
... d = ast.FunctionDef(name='f', args=[], body=[ast.Pass()])
... for i in range(N): d = ast.FunctionDef(name='f', args=[], body=[d])
... m = ast.fix_missing_locations(ast.Module([d]))
... 
... compile(m, "", "exec")
... 
<code object <module> at 0x103baa040, file "", line 1>

```

Reply:
> For consistency, how about pushing a fake context for `with` and try-except here and dropping the check in `compiler.c`?
> To avoid breaking existing code, we'll need to increase `CO_MAXBLOCKS` to, say, 32.

This currently works, but it will stop working if we do that:

>>> d = "def f():"
```suggestion
    if (state->cf_finally_next == CO_MAXBLOCKS) {

>>> N=50; src = "\n".join([' '*i + d for i in range(N)] + [' '*N+'pass'])
>>> compile(src, "", "exec")
<code object <module> at 0x10193c240, file "", line 1>
```

Reply:
Bottom line is I don't think we should use `CO_MAXBLOCKS` in ast_opt.c at all, because we're counting something different that what the compiler is.

Reply:
> > There is no point in being able to parse code that we can't then compile.
> 
> There is no problem compiling a lot of nested functions. A function is not a static block for the compiler.

True, but that doesn't make the above statement false.


It is odd that functions declarations can be nested 100 deep, but other statements only 20.
A single nesting limit makes more sense to me, although this is probably getting out of scope for this PR.

If we have a single nesting limit, it makes senses to enforce those limits in the parser and to simplify the compiler, rather than creating an AST that we then fail to compile.

Reply:
You would still need to check in the compiler in case the ast was constructed by someone other than our parser.

Reply:
Good point.